src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/service/OperationQueueService.java

```diff
@@ -0,0 +1,135 @@
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.reroute.service;
+
+import org.openkilda.messaging.command.CommandData;
+import org.openkilda.messaging.info.InfoData;
+import org.openkilda.wfm.topology.reroute.bolts.OperationQueueCarrier;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Collection;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class OperationQueueService {
+    private final OperationQueueCarrier carrier;
+
+    private final Map<String, FlowQueueData> flowCommands = new HashMap<>();
+
+    public OperationQueueService(OperationQueueCarrier carrier) {
+        this.carrier = carrier;
+    }
+
+    /**
+     * Add command to the front of the queue.
+     */
+    public void addFirst(String flowId, String correlationId, CommandData data) {
+        log.info("Flow command {} has been added to the queue for flow {} with correlation id {}",
+                data, flowId, correlationId);
+        FlowQueueData queueData = getFlowQueueData(flowId);
+        queueData.getQueue().addFirst(new OperationData(correlationId, data));
+        processQueue(queueData);
+    }
+
+    /**
+     * Add command to the end of the queue.
+     */
+    public void addLast(String flowId, String correlationId, CommandData data) {
+        log.info("Flow command {} has been added to the queue for flow {} with correlation id {}",
+                data, flowId, correlationId);
+        FlowQueueData queueData = getFlowQueueData(flowId);
+        queueData.getQueue().addLast(new OperationData(correlationId, data));
+        processQueue(queueData);
+    }
+
+    /**
+     * Handle timeout.
+     */
+    public void handleTimeout(String correlationId) {
+        log.warn("Operation with correlationId {} has timed out", correlationId);
+        Collection<FlowQueueData> foundQueueData = flowCommands.values().stream()
+                .filter(flowQueueData -> correlationId.equals(flowQueueData.getTaskInProgress()))
+                .collect(Collectors.toList());
+
+        if (foundQueueData.isEmpty()) {
+            log.warn("No queue data with correlationId {} found. Timeout event skipped.", correlationId);
```

Comment:
Why warn? Ther is normal to have no such records due to race condition on timeout remove request.

Reply:
Fixed.